documentation/docs/configuration.md
---
# !!! DO NOT MODIFY !!!
# This file is auto-generated by the gendoc tool based on the source code of the app.
description: This section describes the configuration parameters and their types for WASP.
keywords:
- IOTA Node 
- Hornet Node
- WASP Node
- Smart Contracts
- L2
- Configuration
- JSON
- Customize
- Config
- reference
---


# Core Configuration

WASP uses a JSON standard format as a config file. If you are unsure about JSON syntax, you can find more information in the [official JSON specs](https://www.json.org).

You can change the path of the config file by using the `-c` or `--config` argument while executing `wasp` executable.

For example:
```bash
wasp -c config_defaults.json
```

You can always get the most up-to-date description of the config parameters by running:

```bash
wasp -h --full
```

## <a id="app"></a> 1. Application

| Name                      | Description                                            | Type    | Default value |
| ------------------------- | ------------------------------------------------------ | ------- | ------------- |
| checkForUpdates           | Whether to check for updates of the application or not | boolean | true          |
| [shutdown](#app_shutdown) | Configuration for shutdown                             | object  |               |

### <a id="app_shutdown"></a> Shutdown

| Name                     | Description                                                                                            | Type   | Default value |
| ------------------------ | ------------------------------------------------------------------------------------------------------ | ------ | ------------- |
| stopGracePeriod          | The maximum time to wait for background processes to finish during shutdown before terminating the app | string | "5m"          |
| [log](#app_shutdown_log) | Configuration for Shutdown Log                                                                         | object |               |

### <a id="app_shutdown_log"></a> Shutdown Log

| Name     | Description                                         | Type    | Default value  |
| -------- | --------------------------------------------------- | ------- | -------------- |
| enabled  | Whether to store self-shutdown events to a log file | boolean | true           |
| filePath | The file path to the self-shutdown log              | string  | "shutdown.log" |

Example:

```json
  {
    "app": {
      "checkForUpdates": true,
      "shutdown": {
        "stopGracePeriod": "5m",
        "log": {
          "enabled": true,
          "filePath": "shutdown.log"
        }
      }
    }
  }
```

## <a id="logger"></a> 2. Logger

| Name              | Description                                                                 | Type    | Default value |
| ----------------- | --------------------------------------------------------------------------- | ------- | ------------- |
| level             | The minimum enabled logging level                                           | string  | "info"        |
| disableCaller     | Stops annotating logs with the calling function's file name and line number | boolean | true          |
| disableStacktrace | Disables automatic stacktrace capturing                                     | boolean | false         |
| stacktraceLevel   | The level stacktraces are captured and above                                | string  | "panic"       |
| encoding          | The logger's encoding (options: "json", "console")                          | string  | "console"     |
| outputPaths       | A list of URLs, file paths or stdout/stderr to write logging output to      | array   | stdout        |
| disableEvents     | Prevents log messages from being raced as events                            | boolean | true          |

Example:

```json
  {
    "logger": {
      "level": "info",
      "disableCaller": true,
      "disableStacktrace": false,
      "stacktraceLevel": "panic",
      "encoding": "console",
      "outputPaths": [
        "stdout"
      ],
      "disableEvents": true
    }
  }
```

## <a id="inx"></a> 3. INX

| Name                  | Description                                                                                        | Type   | Default value    |
| --------------------- | -------------------------------------------------------------------------------------------------- | ------ | ---------------- |
| address               | The INX address to which to connect to                                                             | string | "localhost:9029" |
| maxConnectionAttempts | The amount of times the connection to INX will be attempted before it fails (1 attempt per second) | uint   | 30               |
| targetNetworkName     | The network name on which the node should operate on (optional)                                    | string | ""               |

Example:

```json
  {
    "inx": {
      "address": "localhost:9029",
      "maxConnectionAttempts": 30,
      "targetNetworkName": ""
    }
  }
```

## <a id="db"></a> 4. Database

| Name                                     | Description                                                                      | Type    | Default value |
| ---------------------------------------- | -------------------------------------------------------------------------------- | ------- | ------------- |
| engine                                   | The used database engine (rocksdb/mapdb)                                         | string  | "rocksdb"     |
| [consensusJournal](#db_consensusjournal) | Configuration for consensusJournal                                               | object  |               |
| [chainState](#db_chainstate)             | Configuration for chainState                                                     | object  |               |
| debugSkipHealthCheck                     | Ignore the check for corrupted databases (should only be used for debug reasons) | boolean | false         |

### <a id="db_consensusjournal"></a> ConsensusJournal

| Name | Description                                       | Type   | Default value             |
| ---- | ------------------------------------------------- | ------ | ------------------------- |
| path | The path to the consensus journal database folder | string | "waspdb/chains/consensus" |

### <a id="db_chainstate"></a> ChainState

| Name | Description                                  | Type   | Default value        |
| ---- | -------------------------------------------- | ------ | -------------------- |
| path | The path to the chain state databases folder | string | "waspdb/chains/data" |

Example:

```json
  {
    "db": {
      "engine": "rocksdb",
      "consensusJournal": {
        "path": "waspdb/chains/consensus"
      },
      "chainState": {
        "path": "waspdb/chains/data"
      },
      "debugSkipHealthCheck": false
    }
  }
```

## <a id="p2p"></a> 5. P2p

| Name               | Description                                             | Type   | Default value |
| ------------------ | ------------------------------------------------------- | ------ | ------------- |
| identityPrivateKey | Private key used to derive the node identity (optional) | string | ""            |
| [db](#p2p_db)      | Configuration for Database                              | object |               |

### <a id="p2p_db"></a> Database

| Name | Description                  | Type   | Default value     |
| ---- | ---------------------------- | ------ | ----------------- |
| path | The path to the p2p database | string | "waspdb/p2pstore" |

Example:

```json
  {
    "p2p": {
      "identityPrivateKey": "",
      "db": {
        "path": "waspdb/p2pstore"
      }
    }
  }
```

## <a id="registry"></a> 6. Registry

| Name                                   | Description                    | Type   | Default value |
| -------------------------------------- | ------------------------------ | ------ | ------------- |
| [chains](#registry_chains)             | Configuration for chains       | object |               |
| [dkShares](#registry_dkshares)         | Configuration for dkShares     | object |               |
| [trustedPeers](#registry_trustedpeers) | Configuration for trustedPeers | object |               |

### <a id="registry_chains"></a> Chains

| Name     | Description                         | Type   | Default value                |
| -------- | ----------------------------------- | ------ | ---------------------------- |
| filePath | The path to the chain registry file | string | "waspdb/chain_registry.json" |

### <a id="registry_dkshares"></a> DkShares

| Name     | Description                                          | Type   | Default value          |
| -------- | ---------------------------------------------------- | ------ | ---------------------- |
| filePath | The path to the distributed key shares registry file | string | "waspdb/dkshares.json" |

### <a id="registry_trustedpeers"></a> TrustedPeers

| Name     | Description                                 | Type   | Default value               |
| -------- | ------------------------------------------- | ------ | --------------------------- |
| filePath | The path to the trusted peers registry file | string | "waspdb/trusted_peers.json" |

Example:

```json
  {
    "registry": {
      "chains": {
        "filePath": "waspdb/chain_registry.json"
      },
      "dkShares": {
        "filePath": "waspdb/dkshares.json"
      },
      "trustedPeers": {
        "filePath": "waspdb/trusted_peers.json"
      }
    }
  }
```

## <a id="peering"></a> 7. Peering

| Name  | Description                                          | Type   | Default value    |
| ----- | ---------------------------------------------------- | ------ | ---------------- |
| netID | Node host address as it is recognized by other peers | string | "127.0.0.1:4000" |
| port  | Port for Wasp committee connection/peering           | int    | 4000             |

Example:

```json
  {
    "peering": {
      "netID": "127.0.0.1:4000",
      "port": 4000
    }
  }
```

## <a id="chains"></a> 8. Chains

| Name                             | Description                                                          | Type    | Default value |
| -------------------------------- | -------------------------------------------------------------------- | ------- | ------------- |
| broadcastUpToNPeers              | Number of peers an offledger request is broadcasted to               | int     | 2             |
| broadcastInterval                | Time between re-broadcast of offledger requests                      | string  | "5s"          |
| apiCacheTTL                      | Time to keep processed offledger requests in api cache               | string  | "5m"          |
| pullMissingRequestsFromCommittee | Whether or not to pull missing requests from other committee members | boolean | true          |

Example:

```json
  {
    "chains": {
      "broadcastUpToNPeers": 2,
      "broadcastInterval": "5s",
      "apiCacheTTL": "5m",
      "pullMissingRequestsFromCommittee": true
    }
  }
```

## <a id="rawblocks"></a> 9. Raw Blocks

| Name      | Description                              | Type    | Default value |
| --------- | ---------------------------------------- | ------- | ------------- |
| enabled   | Whether the raw blocks plugin is enabled | boolean | false         |
| directory | The raw blocks path                      | string  | "blocks"      |

Example:

```json
  {
    "rawBlocks": {
      "enabled": false,
      "directory": "blocks"
    }
  }
```

## <a id="profiling"></a> 10. Profiling

| Name        | Description                                       | Type    | Default value    |
| ----------- | ------------------------------------------------- | ------- | ---------------- |
| enabled     | Whether the profiling plugin is enabled           | boolean | false            |
| bindAddress | The bind address on which the profiler listens on | string  | "localhost:6060" |

Example:

```json
  {
    "profiling": {
      "enabled": false,
      "bindAddress": "localhost:6060"
    }
  }
```

## <a id="wal"></a> 11. Write-Ahead Logging

| Name      | Description                       | Type    | Default value |
| --------- | --------------------------------- | ------- | ------------- |
| enabled   | Whether the WAL plugin is enabled | boolean | true          |
| directory | Path to logs folder               | string  | "wal"         |

Example:

```json
  {
    "wal": {
      "enabled": true,
      "directory": "wal"
    }
  }
```

## <a id="metrics"></a> 12. Metrics

| Name        | Description                             | Type    | Default value    |
| ----------- | --------------------------------------- | ------- | ---------------- |
| enabled     | Whether the metrics plugin is enabled   | boolean | true             |
| bindAddress | The bind address for the node dashboard | string  | "127.0.0.1:2112" |

Example:

```json
  {
    "metrics": {
      "enabled": true,
      "bindAddress": "127.0.0.1:2112"
    }
  }
```

## <a id="webapi"></a> 13. Web API

| Name                      | Description                                              | Type    | Default value    |
| ------------------------- | -------------------------------------------------------- | ------- | ---------------- |
| enabled                   | Whether the web api plugin is enabled                    | boolean | true             |
| nodeOwnerAddresses        | Defines a list of node owner addresses (bech32)          | array   |                  |
| bindAddress               | The bind address for the node web api                    | string  | "127.0.0.1:9090" |
| debugRequestLoggerEnabled | Whether the debug logging for requests should be enabled | boolean | false            |
| [auth](#webapi_auth)      | Configuration for auth                                   | object  |                  |

### <a id="webapi_auth"></a> Auth

| Name                        | Description                            | Type   | Default value |
| --------------------------- | -------------------------------------- | ------ | ------------- |
| scheme                      | Selects which authentication to choose | string | "jwt"         |
| [jwt](#webapi_auth_jwt)     | Configuration for JWT Auth             | object |               |
| [basic](#webapi_auth_basic) | Configuration for Basic Auth           | object |               |
| [ip](#webapi_auth_ip)       | Configuration for IP-based Auth        | object |               |

### <a id="webapi_auth_jwt"></a> JWT Auth

| Name     | Description        | Type   | Default value |
| -------- | ------------------ | ------ | ------------- |
| duration | Jwt token lifetime | string | "24h"         |

### <a id="webapi_auth_basic"></a> Basic Auth

| Name     | Description                                     | Type   | Default value |
| -------- | ----------------------------------------------- | ------ | ------------- |
| username | The username which grants access to the service | string | "wasp"        |

### <a id="webapi_auth_ip"></a> IP-based Auth

| Name      | Description                                          | Type  | Default value |
| --------- | ---------------------------------------------------- | ----- | ------------- |
| whitelist | A list of ips that are allowed to access the service | array | 127.0.0.1     |

Example:

```json
  {
    "webapi": {
      "enabled": true,
      "nodeOwnerAddresses": [],
      "bindAddress": "127.0.0.1:9090",
      "debugRequestLoggerEnabled": false,
      "auth": {
        "scheme": "jwt",
        "jwt": {
          "duration": "24h"
        },
        "basic": {
          "username": "wasp"
        },
        "ip": {
          "whitelist": [
            "127.0.0.1"
          ]
        }
      }
    }
  }
```

## <a id="nanomsg"></a> 14. nanomsg

| Name    | Description                              | Type    | Default value |
| ------- | ---------------------------------------- | ------- | ------------- |
| enabled | Whether the publisher plugin is enabled  | boolean | true          |
| port    | The port for the nanomsg event publisher | int     | 5550          |

Example:

```json
  {
    "nanomsg": {
      "enabled": true,
      "port": 5550
    }
  }
```

## <a id="dashboard"></a> 15. Dashboard

| Name                      | Description                                              | Type    | Default value    |
| ------------------------- | -------------------------------------------------------- | ------- | ---------------- |
| enabled                   | Whether the dashboard plugin is enabled                  | boolean | true             |
| bindAddress               | The bind address for the node dashboard                  | string  | "127.0.0.1:7000" |
| exploreAddressURL         | URL to add as href to addresses in the dashboard         | string  | ""               |
| debugRequestLoggerEnabled | Whether the debug logging for requests should be enabled | boolean | false            |
| [auth](#dashboard_auth)   | Configuration for auth                                   | object  |                  |

### <a id="dashboard_auth"></a> Auth

| Name                           | Description                            | Type   | Default value |
| ------------------------------ | -------------------------------------- | ------ | ------------- |
| scheme                         | Selects which authentication to choose | string | "basic"       |
| [jwt](#dashboard_auth_jwt)     | Configuration for JWT Auth             | object |               |
| [basic](#dashboard_auth_basic) | Configuration for Basic Auth           | object |               |
| [ip](#dashboard_auth_ip)       | Configuration for IP-based Auth        | object |               |

### <a id="dashboard_auth_jwt"></a> JWT Auth

| Name     | Description        | Type   | Default value |
| -------- | ------------------ | ------ | ------------- |
| duration | Jwt token lifetime | string | "24h"         |

### <a id="dashboard_auth_basic"></a> Basic Auth

| Name     | Description                                     | Type   | Default value |
| -------- | ----------------------------------------------- | ------ | ------------- |
| username | The username which grants access to the service | string | "wasp"        |

### <a id="dashboard_auth_ip"></a> IP-based Auth

| Name      | Description                                          | Type  | Default value |
| --------- | ---------------------------------------------------- | ----- | ------------- |
| whitelist | A list of ips that are allowed to access the service | array | 127.0.0.1     |

Example:

```json
  {
    "dashboard": {
      "enabled": true,
      "bindAddress": "127.0.0.1:7000",
      "exploreAddressURL": "",
      "debugRequestLoggerEnabled": false,
      "auth": {
        "scheme": "basic",
        "jwt": {
          "duration": "24h"
        },
        "basic": {
          "username": "wasp"
        },
        "ip": {
          "whitelist": [
            "127.0.0.1"
          ]
        }
      }
    }
  }
```



documentation/docs/guide/development_tools/docker_preconfigured.md
---
description: How to run the preconfigured Docker setup.
image: /img/logo/WASP_logo_dark.png
keywords:

- smart contracts
- validator node
- Docker
- image
- build
- configure
- arguments
- Hornet
- how to

---

# Preconfigured Development Docker Setup

This page describes how you can use the preconfigured developer Docker setup.

:::note
This setup is intended for **local** development only. You will have your own private network/tangle.
:::

## Introduction

To diminish the time spent on configuration and research, we have created a docker-compose setup that ships a
pre-configured Wasp node with a 
[Hornet-based private tangle](https://wiki.iota.org/shimmer/hornet/how_tos/private_tangle), ready to run out of the box.

## Requirements

* [Docker](https://www.docker.com/)
* [Docker Compose](https://docs.docker.com/compose/)
* [Git](https://git-scm.com/)

## Quick Start Guide

1. Checkout the project:

    ```shell
    git clone https://github.com/iotaledger/wasp.git
    ```

2. Move into the project directory and check out the `develop` branch:

    ```shell
   cd wasp
   git checkout develop
    ```

3. Move into the `tools/local-setup` folder:

    ```shell
     cd tools/local-setup
    ```

4. Run the following command to start the setup.

    ```shell
    docker-compose up
    ```

This command will initialize Hornet and create a fresh image of the checked-out Wasp code.

If you modify the branch, docker-compose will include your modifications in the Wasp image.

:::note Default Ports

All Wasp ports will bind to 127.0.0.1 by default.
If you want to expose the ports to the outside world, run `HOST=0.0.0.0 docker-compose up`.

:::

## Wasp-CLI Configuration

All of the `wasp-cli` ports are locally available. You should use this `wasp-cli.json` configuration:

```json
{
  "l1": {
    "apiaddress": "http://localhost:14265",
    "faucetaddress": "http://localhost:8091"
  },
  "wasp": {
    "0": {
      "api": "127.0.0.1:9090",
      "nanomsg": "127.0.0.1:5550",
      "peering": "127.0.0.1:4000"
    }
  }
}
```

Run `wasp-cli init` to generate a seed, and you are ready to go.

See [Configuring wasp-cli](../chains_and_nodes/wasp-cli) for further information.

## Usage

By default, Wasp will allow any connection coming from `wasp-cli`. This is fine for development purposes, but please
ensure you don’t run it on a publicly available server or create matching firewall filter rules if you do so.

Other than that, everything should simply work as expected. Faucet requests will be handled. You will be able
to deploy and run smart contracts. All useful [ports](#reachable-ports) are available to the local machine.

### Start

To start the setup, run:

```bash
docker-compose up
```

During the startup, you might see a few failed restarts of Wasp with the message:
`panic: error getting node event client: mqtt plugin not available on the current node.`

This is normal, as Wasp starts faster than Hornet. Wasp retries the connection until it succeeds.

### Stop

You can shut down the setup by running the following command in a new terminal:

```shell
docker-compose down
```

You can also press `Ctrl-C` to shut down the setup, but **don't press it twice to force it**. Otherwise, you can corrupt
the Hornet database.

### Reset

To shut down the nodes and remove all databases, run:

```shell
docker-compose down --volumes
```

### Recreation

If you made changes to the Wasp code and want to use it inside the setup, you need to recreate the Wasp image by
running:

```bash
docker-compose build
```

## Reachable Ports

The nodes will then be reachable under these ports:

### Wasp

#### API

[http://localhost:9090]( http://localhost:9090)

#### Dashboard

[http://localhost:7000](http://localhost:7000) (**Username**: wasp  **Password**: wasp)

#### Nanomsg

[tcp://localhost:5550](tcp://localhost:5550)

### Hornet

#### API

[http://localhost:14265](http://localhost:14265)

#### Faucet

[http://localhost:8091](http://localhost:8091)

#### Dashboard

[http://localhost:8081](http://localhost:8081) (**Username**: admin **Password**: admin)


tools/wasp-cli/config/config.go
package config

import (
	"fmt"
	"time"

	"github.com/spf13/cobra"
	"github.com/spf13/viper"

	"github.com/iotaledger/wasp/client"
	"github.com/iotaledger/wasp/packages/nodeconn"
	"github.com/iotaledger/wasp/packages/parameters"
	"github.com/iotaledger/wasp/packages/testutil/privtangle/privtangledefaults"
	"github.com/iotaledger/wasp/tools/wasp-cli/log"
)

var (
	ConfigPath        string
	WaitForCompletion bool
)

var configSetCmd = &cobra.Command{
	Use:   "set <key> <value>",
	Short: "Set a configuration value",
	Args:  cobra.ExactArgs(2),
	Run: func(cmd *cobra.Command, args []string) {
		v := args[1]
		switch v {
		case "true":
			Set(args[0], true)
		case "false":
			Set(args[0], false)
		default:
			Set(args[0], v)
		}
	},
}

var refreshL1ParamsCmd = &cobra.Command{
	Use:   "refresh-l1-params",
	Short: "Refresh L1 params from node",
	Args:  cobra.NoArgs,
	Run: func(cmd *cobra.Command, args []string) {
		refreshL1ParamsFromNode()
	},
}

const (
	HostKindAPI     = "api"
	HostKindPeering = "peering"
	HostKindNanomsg = "nanomsg"

	l1ParamsKey          = "l1.params"
	l1ParamsTimestampKey = "l1.timestamp"
	l1ParamsExpiration   = 24 * time.Hour
)

func Init(rootCmd *cobra.Command) {
	rootCmd.PersistentFlags().StringVarP(&ConfigPath, "config", "c", "wasp-cli.json", "path to wasp-cli.json")
	rootCmd.PersistentFlags().BoolVarP(&WaitForCompletion, "wait", "w", true, "wait for request completion")

	rootCmd.AddCommand(configSetCmd)
	rootCmd.AddCommand(checkVersionsCmd)
	rootCmd.AddCommand(refreshL1ParamsCmd)

	// The first time parameters.L1() is called, it will be initialized with this function
	parameters.InitL1Lazy(func() {
		if l1ParamsExpired() {
			refreshL1ParamsFromNode()
		} else {
			loadL1ParamsFromConfig()
		}
	})
}

func l1ParamsExpired() bool {
	if viper.Get(l1ParamsKey) == nil {
		return true
	}

	return viper.GetTime(l1ParamsTimestampKey).Add(l1ParamsExpiration).Before(time.Now())
}

func refreshL1ParamsFromNode() {
	if log.VerboseFlag {
		log.Printf("Getting L1 params from node at %s...\n", L1APIAddress())
	}
	L1Client() // this will call parameters.InitL1()
	Set(l1ParamsKey, parameters.L1())
	Set(l1ParamsTimestampKey, time.Now())
}

func loadL1ParamsFromConfig() {
	// read L1 params from config file
	var params *parameters.L1Params
	err := viper.UnmarshalKey("l1.params", &params)
	log.Check(err)
	parameters.InitL1(params)
}

func Read() {
	viper.SetConfigFile(ConfigPath)
	_ = viper.ReadInConfig()
}

func L1APIAddress() string {
	host := viper.GetString("l1.apiAddress")
	if host != "" {
		return host
	}

	return fmt.Sprintf(
		"%s:%d",
		privtangledefaults.Host,
		privtangledefaults.BasePort+privtangledefaults.NodePortOffsetRestAPI,
	)
}

func L1FaucetAddress() string {
	address := viper.GetString("l1.faucetAddress")
	if address != "" {
		return address
	}

	return fmt.Sprintf(
		"%s:%d",
		privtangledefaults.Host,
		privtangledefaults.BasePort+privtangledefaults.NodePortOffsetFaucet,
	)
}

func L1Client() nodeconn.L1Client {
	log.Verbosef("using L1 API %s\n", L1APIAddress())

	return nodeconn.NewL1Client(
		nodeconn.L1Config{
			APIAddress:    L1APIAddress(),
			FaucetAddress: L1FaucetAddress(),
		},
		log.HiveLogger(),
	)
}

func GetToken() string {
	return viper.GetString("authentication.token")
}

func SetToken(token string) {
	Set("authentication.token", token)
}

func WaspClient(i ...int) *client.WaspClient {
	// TODO: add authentication for /adm
	log.Verbosef("using Wasp host %s\n", WaspAPI())
	L1Client() // this will fill parameters.L1() with data from the L1 node

	return client.NewWaspClient(WaspAPI(i...)).WithToken(GetToken())
}

func WaspAPI(i ...int) string {
	index := 0
	if len(i) > 0 {
		index = i[0]
	}
	r := viper.GetString("wasp." + HostKindAPI)
	if r != "" {
		return r
	}

	return committeeHost(HostKindAPI, index)
}

func WaspNanomsg(i ...int) string {
	index := 0
	if len(i) > 0 {
		index = i[0]
	}
	r := viper.GetString("wasp." + HostKindNanomsg)
	if r != "" {
		return r
	}

	return committeeHost(HostKindNanomsg, index)
}

func FindNodeBy(kind, v string) int {
	for i := 0; i < 100; i++ {
		if committeeHost(kind, i) == v {
			return i
		}
	}
	log.Fatalf("Cannot find node with %q = %q in configuration", kind, v)

	return 0
}

func CommitteeAPI(indices []int) []string {
	return committee(HostKindAPI, indices)
}

func CommitteePeering(indices []int) []string {
	return committee(HostKindPeering, indices)
}

func CommitteeNanomsg(indices []int) []string {
	return committee(HostKindNanomsg, indices)
}

func committee(kind string, indices []int) []string {
	hosts := make([]string, 0)
	for _, i := range indices {
		hosts = append(hosts, committeeHost(kind, i))
	}

	return hosts
}

func committeeConfigVar(kind string, i int) string {
	return fmt.Sprintf("wasp.%d.%s", i, kind)
}

func CommitteeAPIConfigVar(i int) string {
	return committeeConfigVar(HostKindAPI, i)
}

func CommitteePeeringConfigVar(i int) string {
	return committeeConfigVar(HostKindPeering, i)
}

func CommitteeNanomsgConfigVar(i int) string {
	return committeeConfigVar(HostKindNanomsg, i)
}

func committeeHost(kind string, i int) string {
	r := viper.GetString(committeeConfigVar(kind, i))
	if r != "" {
		return r
	}
	defaultPort := defaultWaspPort(kind, i)

	return fmt.Sprintf("127.0.0.1:%d", defaultPort)
}

func totalNumberOfWaspNodes() int {
	return len(viper.Sub("wasp").AllSettings())
}

func defaultWaspPort(kind string, i int) int {
	switch kind {
	case HostKindNanomsg:
		return 5550 + i
	case HostKindPeering:
		return 4000 + i
	case HostKindAPI:
		return 9090 + i
	}
	panic(fmt.Sprintf("no handler for kind %s", kind))
}

func Set(key string, value interface{}) {
	viper.Set(key, value)
	log.Check(viper.WriteConfig())
}
